### ISSUE TITLE
Error with JitTestCase

### ISSUE THREAD

**Issue by anijain2305** [pytorch/torchdynamo#376]
Repro

~~~
import torch
import torchdynamo
from torch.testing._internal.jit_utils import JitTestCase

class TestTracer(JitTestCase):
    def test(self):
        class Foo(torch.nn.Module):
            def __init__(self):
                super(Foo, self).__init__()
                self.a = 3

            @torch.jit.export
            def __getstate__(self):
                return (3, self.training)

            @torch.jit.export
            def __setstate__(self, state):
                self.a = state[0]
                self.training = state[1]

            def forward(self, x):
                return x + self.a

        f = Foo()

        return torch.jit.trace(f, (torch.rand(3, 4),))


obj = TestTracer()
eager = obj.test()
with torchdynamo.optimize("eager"):
    dynamo = obj.test()

print(eager)
print(dynamo)


~~~

Fails with

~~~
E           RuntimeError: The field 'generation' was left uninitialized after '__setstate__', but expected a value of type 'int'
~~~

**NOTE** that it fails only when `pytest` is called  - `pytest test_setstate.py` 
If you run with `python test_setstate.py`, it passes. 
I have seen around 25 failures with this error message.

cc @mlazos @jansel @voznesenskym

**Comment by mlazos** [comment]
Discussed a bit offline - what I'm thinking here is that the object is getting serialized in a context where we have patched `__init__` and `__set_state__` and thus has a generation tag and so the serialized data has this tag. Then when it is deserialized, it is deserialized in a context where the patch hasn't been applied, so this assertion is triggered when there is a field in the serialized data which isn't set by `__set_state__`.

Two thoughts:
We might have to look at what the `@torch.jit.export` tag does and make sure that when we patch `__set_state__` this decorator is applied.

Another possible solution is if we just ignore this check, but that feels a little dirtier. The reason we can ignore is that if we are outside a dynamo context (where the patch hasn't been applied) then we don't care about the generation anyway, and once we enter a context, the module will be considered non-dynamic, which is the same result we would get if we always have `__set_state__` correctly setting it.

**Comment by anijain2305** [comment]
Closed with #423
